File: frontend/src/stores/warehouse.ts
import { defineStore } from 'pinia'
import { warehouseApi } from '@/api/warehouse'

export interface Warehouse {
  warehouse_id: string
  name: string
  last_scan_time: string | null
  container_count: number
  total_items: number
  material_count: number
  group_id: string | null
  organization_id: string | null
  teleport_cmd: string | null
  logistics_teleport_cmd: string | null
}

export interface WarehouseZone {
  zone_id: string
  warehouse_fk: string
  name: string
  range_min_x: number
  range_min_y: number
  range_min_z: number
  range_max_x: number
  range_max_y: number
  range_max_z: number
  aisle_lines: any[]
  created_at: string | null
}

export interface MaterialItem {
  item_id: string
  display_name: string
  count: number
}

export const useWarehouseStore = defineStore('warehouse', {
  state: () => ({
    warehouses: [] as Warehouse[],
    currentWarehouse: null as Warehouse | null,
    materials: [] as MaterialItem[],
    materialTotal: 0,
    loading: false,
    // 扫描状态（Socket.IO 实时推送）
    scanStatus: 'idle' as string, // idle | queued | scanning | paused | finished | cancelled | failed
    scanProgress: 0,
    scanScanned: 0,
    scanTotal: 0,
    scanItems: 0,
    scanSpeed: 0,
    scanEta: null as number | null,
    scanCurrentPos: null as { x: number; y: number; z: number } | null,
    scanQueue: [] as any[], // 扫描队列（scan_queue_update）
  }),
  actions: {
    async fetchWarehouses() {
      this.loading = true
      try {
        const { data } = await warehouseApi.list()
        this.warehouses = data
      } finally {
        this.loading = false
      }
    },
    async fetchWarehouse(id: string) {
      const { data } = await warehouseApi.get(id)
      this.currentWarehouse = data
      return data
    },
    async createWarehouse(name: string, teleportCmd?: string) {
      const { data } = await warehouseApi.create({ name, teleport_cmd: teleportCmd || undefined })
      this.warehouses.push(data)
      return data
    },
    async updateWarehouse(id: string, data: any) {
      const res = await warehouseApi.update(id, data)
      this.currentWarehouse = res.data
      return res.data
    },
    async deleteWarehouse(id: string) {
      await warehouseApi.delete(id)
      this.warehouses = this.warehouses.filter(w => w.warehouse_id !== id)
    },
    async fetchMaterials(id: string) {
      const { data } = await warehouseApi.getMaterials(id)
      this.materials = data.items ?? []
      this.materialTotal = data.total ?? 0
      return data
    },
    async fetchScanStatus(id: string) {
      const { data } = await warehouseApi.getScanStatus(id)
      this.scanStatus = data.status || 'idle'
      this.scanProgress = data.progress || 0
      this.scanScanned = data.scanned_containers || 0
      this.scanTotal = data.total_containers || 0
      this.scanItems = data.items_scanned || 0
      return data
    },
    async fetchScanQueue() {
      try {
        const { data } = await warehouseApi.getScanQueue()
        this.scanQueue = data?.items ?? []
      } catch { /* ignore */ }
    },
    setScanProgress(payload: any) {
      this.scanProgress = payload?.progress ?? this.scanProgress
      this.scanScanned = payload?.scanned ?? this.scanScanned
      this.scanTotal = payload?.total ?? this.scanTotal
      this.scanItems = payload?.items_scanned ?? this.scanItems
      this.scanSpeed = payload?.speed ?? 0
      this.scanEta = payload?.eta_seconds ?? null
      if (payload?.current_pos) this.scanCurrentPos = payload.current_pos
    },
    setScanQueue(items: any[]) {
      this.scanQueue = items ?? []
    },
  },
})

File: vmtools-next/src/vmtools_next/data/models/warehouse.py
"""Warehouse-related ORM models (migrated from vmtools-backend).

Tables: warehouses, material_items, container_items, scan_status, storage_zones
"""
from __future__ import annotations

from datetime import datetime, timezone

from sqlalchemy import (
    Column, String, Integer, BigInteger, Float, DateTime, ForeignKey,
    UniqueConstraint, Text, Boolean, Index,
)
from sqlalchemy.orm import relationship

from vmtools_next.data.db import Base


class WarehouseModel(Base):
    """A registered warehouse (scanned region)."""

    __tablename__ = "warehouses"

    warehouse_id = Column(String, primary_key=True)
    name = Column(String, default="")
    last_scan_time = Column(DateTime, nullable=True)
    container_count = Column(BigInteger, default=0)
    total_items = Column(BigInteger, default=0)
    aisle_lines = Column(Text, default="[]")
    group_id = Column(String, ForeignKey("warehouse_groups.id"), nullable=True, index=True)
    organization_id = Column(String, ForeignKey("organizations.id"), nullable=True, index=True)
    logistics_enabled = Column(Boolean, default=False, nullable=False)
    logistics_teleport_cmd = Column(String, nullable=True)

    materials = relationship("MaterialItemModel", back_populates="warehouse", cascade="all, delete-orphan")
    containers = relationship("ContainerItemModel", back_populates="warehouse", cascade="all, delete-orphan")
    scan_status = relationship("ScanStatusModel", back_populates="warehouse", uselist=False, cascade="all, delete-orphan")
    zones = relationship("StorageZoneModel", back_populates="warehouse", cascade="all, delete-orphan", order_by="StorageZoneModel.created_at")


class MaterialItemModel(Base):
    """Aggregated material totals for a warehouse."""

    __tablename__ = "material_items"

    id = Column(Integer, primary_key=True, autoincrement=True)
    warehouse_fk = Column(String, ForeignKey("warehouses.warehouse_id"), index=True)
    item_id = Column(String, index=True)
    display_name = Column(String)
    count = Column(BigInteger)
    nbt_hash = Column(String, nullable=True)

    __table_args__ = (
        Index("idx_material_items_wh_item", "warehouse_fk", "item_id"),
    )

    warehouse = relationship("WarehouseModel", back_populates="materials")


class ContainerItemModel(Base):
    """Per-container snapshot — one row per container position."""

    __tablename__ = "container_items"

    id = Column(Integer, primary_key=True, autoincrement=True)
    warehouse_fk = Column(String, ForeignKey("warehouses.warehouse_id"), index=True)
    container_x = Column(Integer, nullable=False)
    container_y = Column(Integer, nullable=False)
    container_z = Column(Integer, nullable=False)
    item_id = Column(String, index=True)
    item_name_zh = Column(String)
    count = Column(BigInteger, default=0)

    __table_args__ = (
        UniqueConstraint("warehouse_fk", "container_x", "container_y", "container_z", name="uq_container_pos"),
        Index("idx_container_items_wh_y_z", "warehouse_fk", "container_y", "container_z"),
        Index("idx_container_items_wh_item", "warehouse_fk", "item_id"),
    )

    warehouse = relationship("WarehouseModel", back_populates="containers")


class ScanStatusModel(Base):
    """Live scan progress for a warehouse."""

    __tablename__ = "scan_status"

    warehouse_fk = Column(String, ForeignKey("warehouses.warehouse_id"), primary_key=True)
    status = Column(String, default="idle")  # idle|scanning|paused|finished|cancelled
    progress = Column(Float, default=0.0)
    current_pos = Column(String, nullable=True)
    total_containers = Column(BigInteger, default=0)
    scanned_containers = Column(BigInteger, default=0)
    failed_containers = Column(BigInteger, default=0)

    warehouse = relationship("WarehouseModel", back_populates="scan_status")


class StorageZoneModel(Base):
    """A storage zone within a warehouse, with its own range and aisle lines."""

    __tablename__ = "storage_zones"

    zone_id = Column(String, primary_key=True)
    warehouse_fk = Column(String, ForeignKey("warehouses.warehouse_id"), index=True)
    name = Column(String, default="")
    range_min_x = Column(Integer, default=0)
    range_min_y = Column(Integer, default=0)
    range_min_z = Column(Integer, default=0)
    range_max_x = Column(Integer, default=0)
    range_max_y = Column(Integer, default=0)
    range_max_z = Column(Integer, default=0)
    aisle_lines = Column(Text, default="[]")
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))

    warehouse = relationship("WarehouseModel", back_populates="zones")
